### DISCUSSION TITLE
Troubleshooting help wanted: zellij broke bash completion on some setups?

### DISCUSSION THREAD

**Discussion by toymil** [zellij-org/zellij#3035] [Q&A] [answered]
I don't think this is a bug of zellij, rather a bug related to zellij?  However I don't know how to reproduce the bug outside of the 2 problematic systems, I am looking for insights and suggestions to debug more.

Either way I do like to solve it since it's making the shell unusable.

## system info

- *local pc*
  - installation detail
    - pc
    - windows 10 19044.3803
      - wsl 2.0.14.0
      - debian testing (13, trixie)
  - package versions
    - `bash`: 5.2.21-2
    - `bash-completion`: 1:2.11-8
    - `zellij`: 0.39.2 (`cargo install`)
- *remote 1*
  - installation detail
    - home server
    - debian 12.4, standard TUI install from iso image
  - package versions
    - `bash`: 5.2.15-2+b2
    - `bash-completion`: 1:2.11-6
    - `zellij`: 0.39.2 (`cargo install`)
- *remote 2 & 3*
  - installation detail
    - vps (from 2 different providers)
    - provider preinstalled debian 10, later upgraded to 11, then 12.4
  - package versions
    - `bash`: 5.2.15-2+b2
    - `bash-completion`: 1:2.11-6
    - `zellij`: 0.39.2 (binary copied from *local pc*)

## expected behavior

- complete command names, directories, executables
- complete command specific arguments, e.g.:
  - `git com` TAB -> `git commit`
  - `systemctl status ` TAB -> `systemctl status <(list of) service name>`
- double TAB at empty prompt does nothing

The following setups work as expected:
- *local pc* bash
- *local pc* bash > ssh *remote 1* bash
- *local pc* bash > ssh *remote 1* bash > zellij
- *local pc* bash > zellij
- *local pc* bash > zellij > ssh *remote 1* bash
- *local pc* bash > zellij > ssh *remote 1* bash > zellij

## problem on *remote 2 and 3*

- at the `0`th argument, everything seems to work, still completes command names, directories, executables
- **does not complete command specific arguments, will only try to suggest dir and file names**, e.g.:
  - `git com` TAB does not complete to `git commit`, and will do nothing if current dir does not contain file or dir starting with `com`
  - `systemctl status ` TAB gives a list of file and dir
- **double TAB at empty prompt will try to list all aliases and commands**

Setups without the problem:
- *local pc* bash > ssh *remote 2 or 3* bash
- *local pc* bash > zellij > ssh *remote 2 or 3* bash
- *local pc* bash > ssh *remote 2* bash > zellij > ssh *remote 3* bash
- *local pc* bash > zellij > ssh *remote 2* bash > zellij > ssh *remote 3* bash

***NOT working setup***:
- *local pc* bash > ssh *remote 2 or 3* bash > zellij

## summary

Problem occurs on *remote 2 and 3* only when zellij is running on the *end system*.  Problem does not occur if zellij is merely wrapping the ssh session (i.e. not running on the *end system*).

I have no idea how zellij could be causing this problem, or how to debug this situation.  Base on the gathered information I can only make 2 hypothesis:
- *remote 2 and 3* system is missing some package or configuration since the base vendor provided installation might not be a standard one.
- ~~the zellij used on *remote 2 and 3* are not compiled on the same debian version, somehow causing the problem.~~
  - binary from *remote 1* does not change anything

**Comment by imsnif** [comment]
Hey, thanks for opening this issue. Zellij does not interact with your system completions (local or remote). This is 99% an environmental issue (having to do with how you run Zellij, how you invoke your shell, where your completions are defined, etc.)

Troubleshooting your system until we find the root cause is unfortunately out of scope for issues in this repository. Issues are reserved to bugs or feature requests in the Zellij software. You are of course welcome to ask for help troubleshooting your system on Discord/Matrix, but otherwise I'm afraid I'm going to have to close this issue.

If you find an issue with Zellij itself and can provide a minimal reproduction, I'd be happy to take a look at it.

Happy new year.

**Comment by toymil** [comment]
@imsnif can you convert this issue to a discussion then?  I am pretty sure if I ask this in the group chat it will just disappear into the stream...

**Reply by imsnif** [reply]
Sure, done.

**Reply by imsnif** [reply]
I also changed the title to make it clearer and help people find it.

**Comment by toymil** [comment]
I think I almost figured it out.  @imsnif one question to help me understand: when I run the `w` command, there will only be 1 entry, no matter how many panel/tab I open in zellij.  What is this and where can I learn more about it?  `tmux` does not have this behavior.

**Comment by toymil** [comment, marked as answer]
On all the working machines, there are 2 places where completion could be loaded:
- `~/.bashrc`
- `/etc/profile` -> `/etc/profile.d/bash_completion.sh`

On *remote 2 and 3* there is only one:
- `/etc/profile` -> `/etc/profile.d/bash_completion.sh`

`/etc/profile` will run once on shell login, but then due to some zellij design it *will not run again after zellij launch* (because any panel/tab inside zellij is not a new shell?), so the shell inside zellij did not invoke `/etc/profile.d/bash_completion.sh` thus did not load shell completion.

All the working machines do not have the problem because `~/.bashrc` does run in each new zellij panel/tab.
